### PR TITLE
Add a switch to defer rendering of components in ComponentTagHelper

### LINKED ISSUE #38400: Investigate not requiring unintuitive requirements for pre-rendering PageTitle
Pre-rendering for PageTitle requires rendering things in a specific order and relies on  a _Layout.cshtml which is not very intuitive. See https://docs.microsoft.com/en-us/aspnet/core/blazor/components/prerendering-and-integration?view=aspnetcore-6.0&pivots=webassembly#prerendering-configuration for all of the steps involved here.

Possible things to look at:

a) Can we do away with requiring a separate `_Layout`?
b) If the first option is difficult, could we produce an error if the `PageTitle` component is being rendered before any other component?

### PR DESCRIPTION
**Pull request by pranavkm**
This allows the HeadOutlet component to appear earlier in the document, but for
rendering to be deferred until MVC starts emitting markup.

The way this is done is by introducing a new parameter to `ComponentTagHelper` / `RenderComponentAsync` which returns a IHtmlContent that defers rendering the component until `IHtmlContent.WriteTo` is called. In practice this looks like:

```razor
<head>
     ....
    <component type="typeof(HeadOutlet)" render-mode="ServerPrerendered" defer="true" />
</head>
<body>
    ....
   <component type="typeof(App)" render-mode="ServerPrerendered" />
</body>
```

This behavior takes advantage of MVC's buffered rendering mechanism. MVC builds a tree of IHtmlContent items and defers calling `WriteTo` on these until it's ready to go to the wire. In our case, the component tag helper rendering `HeadOutlet` executes but puts a placeholder since it is deferred. The tag helper rendering App executes, and updates the SectionContent. Once MVC is ready to turn IHtmlContent to text, the `HeadOutlet` component renders and the title configured by `App` is used. 

The only pre-req is that since rendering of `IHtmlContent` is synchronous, we require deferred components to also finish rendering synchronously. This isn't an issue with `HeadOutlet`.

Fixes https://github.com/dotnet/aspnetcore/issues/38400

### CODE REVIEW

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs [hunk 1]
```diff
@@ -25,7 +25,7 @@ public BlazorServerTemplateTest(ProjectFactoryFixture projectFactory)
 
     public override string ProjectType { get; } = "blazorserver";
 
-    [Theory(Skip = "https://github.com/dotnet/aspnetcore/issues/30761")]
```
Comment (pranavkm) [inline src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs:28]: Will undo - I wanted to verify if the title in the templates worked correctly

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/_Host.cshtml [hunk 2]
```diff
@@ -1,8 +1,33 @@
 ﻿@page "/"
+@using Microsoft.AspNetCore.Components.Web
 @namespace BlazorServerWeb_CSharp.Pages
 @addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
-@{
-    Layout = "_Layout";
-}
 
-<component type="typeof(App)" render-mode="ServerPrerendered" />
+<!DOCTYPE html>
```
Comment (pranavkm) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/_Host.cshtml:6]: Merged `_Host.cshtml` and `_Layout.cshtml` together like pre-6.0 days.

File: src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs [hunk 3]
```diff
@@ -67,6 +92,42 @@ internal sealed class ComponentRenderer : IComponentRenderer
...
+        }
+
+        public void WriteTo(TextWriter writer, HtmlEncoder encoder)
+        {
+            var renderTask = _componentRenderer.RenderComponentCoreAsync(
+                _viewContext, _componentType, _renderMode, _parameterView);
+
+            if (!renderTask.IsCompleted)
+            {
+                throw new InvalidOperationException("Components with deferred rendering must complete synchronously.");
```
Comment (pranavkm) [inline src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs:124]: I'll update this to use a resx.

**Comment by SteveSandersonMS** [comment]
This looks very promising to me. I'm definitely enthusiastic about getting rid of the weird host/layout page split and the nonobvious ordering dependency.

I'm wondering though if it could be simplified further, eliminating the need for the `Defer` parameter and avoiding any limitation about async rendering. I might be missing something essential but it seems like it might be possible:

How expensive is it to use IHtmlContent? Would it be viable for *all* component prerendering to go through this? Imagine if:

1. Every prerendered component wrote its output (asynchronously) to an IHtmlContent.
2. Each time a prerendered root component re-renders, we'd clear the contents of its IHtmlContent and write the new content (again, asynchronously).
3. Finally, when MVC wants to emit the page output to the response, it calls `WriteTo` on all the IHtmlContent instances and they all synchronously flush out their most recent render output

So if rendering `App` causes the `HeadOutlet` to re-render (which it does - that's exactly how this whole thing works in non-prerendering cases), then by the time `App` has finished asynchronously prerendering, we would also have finished asynchronously re-pre-rendering `HeadOutlet`. This is because RenderRootComponentAsync gives a task that only completes when the whole renderer becomes quiescent (at least I think so).

Is there a fatal flaw in this? Is it cheap enough to buffer each prerendered component like this?

**Comment by pranavkm** [comment]
Closing in favor of https://github.com/dotnet/aspnetcore/pull/40512

**Pull request closed without merging**